airflow/migrations/versions/cc1e65623dc7_add_max_tries_column_to_task_instance.py

```diff
@@ -0,0 +1,63 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""add max tries column to task instance
+
+Revision ID: cc1e65623dc7
+Revises: 127d2bf2dfa7
+Create Date: 2017-06-19 16:53:12.851141
+
+"""
+
+# revision identifiers, used by Alembic.
+revision = 'cc1e65623dc7'
+down_revision = '127d2bf2dfa7'
+branch_labels = None
+depends_on = None
+
+from alembic import op
+import sqlalchemy as sa
+from airflow import settings
+from airflow.models import DagBag, TaskInstance
+
+
+def upgrade():
+    op.add_column('task_instance', sa.Column('max_tries', sa.Integer, server_default="-1"))
+    # Check if table task_instance exist before data migration
+    engine = settings.engine
+    if engine.dialect.has_table(engine, 'task_instance'):
+        # Get current session
+        connection = op.get_bind()
+        Session = sa.orm.sessionmaker()
+        session = Session(bind=connection)
+        dagbag = DagBag(settings.DAGS_FOLDER)
+        query = session.query(sa.func.count(TaskInstance.max_tries)).filter(
+            TaskInstance.max_tries == -1
+        )
+        while query.scalar():
+            tis = session.query(TaskInstance).filter(
+                TaskInstance.max_tries == -1
+            ).limit(5000).all()
+            for ti in tis:
+                dag = dagbag.get_dag(ti.dag_id)
+                if not dag or not dag.has_task(ti.task_id):
+                    ti.max_tries = ti.try_number
+                else:
+                    task = dag.get_task(ti.task_id)
+                    ti.max_tries = task.retries
+                session.merge(ti)
+        session.commit()
+
+
+def downgrade():
+    op.drop_column('task_instance', 'max_tries')
```

Comment:
hmm I think try_number shouldn't be modified in this case since the change only affects max_tries, i.e try_number should be unaware of this new column and any logic related to max_tries and try_number shouldn't be reflected here. (Because we only initialize max_tries to equal try_number but have no idea the relationship between max_tries and try_number...yet!)